### PR TITLE
Adding f16 as Dtype

### LINKED ISSUE #423: Add f16/bf16 dtype support for tensors
This can use the `half` crate.

Cuda has many arithmetic functions available for half as seen [here](https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__INTRINSIC__HALF.html#group__CUDA__MATH__INTRINSIC__HALF)

hgemm support depends on https://github.com/coreylowman/cudarc/issues/65

### PR DESCRIPTION
**Pull request by coreylowman**
Resolves #423

### CODE REVIEW

**Comment by coreylowman** [comment]
FYI @opfromthestart @nkoppel If you guys want to work on this, you can open PRs into this feature branch

File: Cargo.toml [hunk 1]
```diff
@@ -46,8 +47,8 @@ indicatif = "0.17.3"
 glob = { version = "0.3.1", optional = true }
 
 [features]
-default = ["std", "fast-alloc", "cpu-par-matmul"]
-nightly = []
+default = ["std", "fast-alloc", "cpu-par-matmul", "test-f16", "nightly"]
```
Comment (nkoppel) [inline Cargo.toml:50]:
```suggestion
default = ["std", "fast-alloc", "cpu-par-matmul", "f16", "nightly"]
```
Testing with f32 should be the default for now.

File: src/lib.rs [hunk 2]
```diff
@@ -275,6 +293,32 @@ pub(crate) mod tests {
         }
     }
 
+    #[cfg(feature = "f16")]
+    impl AssertClose for half::f16 {
+        type Elem = Self;
+        const DEFAULT_TOLERANCE: Self::Elem = half::f16::from_f32_const(1e-3);
```
Comment (coreylowman) [inline src/lib.rs:299]: What should this tolerance be?

File: src/lib.rs [hunk 3]
```diff
@@ -275,6 +293,32 @@ pub(crate) mod tests {
...
+            } else {
+                None
+            }
+        }
+    }
+
+    #[cfg(feature = "bf16")]
+    impl AssertClose for half::bf16 {
+        type Elem = Self;
+        const DEFAULT_TOLERANCE: Self::Elem = half::bf16::from_f32_const(1e-3);
```
Comment (coreylowman) [inline src/lib.rs:312]: What should this tolerance be?

**Comment by coreylowman** [comment]
I'm wondering if we should enforce f16/bf16 tests passing. There's particular tests where I just don't think it'll have the accuracy required to pass the tests, or we need a way to reduce the tolerance even more.

Here are the current failing t ests for f16:
```
losses::tests::test_hard_crossentropy
    nn::batchnorm1d::tests::test_batchnorm1d_2d_forward_mut
    nn::batchnorm2d::tests::test_batchnorm2d_4d_forward_mut
    nn::linear::tests::test_forward_1d
    nn::linear::tests::test_linear_initialize
    nn::residual::tests::test_residual_gradients
    nn::unbiased_linear::tests::test_forward_1d
    optim::adam::tests::test_adam_decoupled_decay
    optim::adam::tests::test_custom_adam_one_params
    optim::adam::tests::test_default_adam_params
    tensor_ops::matmul::tests::test_matmul_vec_normal
    tensor_ops::matmul::tests::test_matmul_vec_transpose
    tensor_ops::matmul::tests::test_small_matmul_mm
    tensor_ops::max_to::tests::test_max_axis_0_2d
    tensor_ops::mean_to::tests::test_mean_axis_0_2d
    tensor_ops::pool2d::tests::test_pool2d_3d_max2d
    tensor_ops::sum_to::tests::test_sum_axes_3d_to_1d
    tensor_ops::sum_to::tests::test_sum_broadcasted
    tensor_ops::upscale2d::tests::test_bilinear_upscale2d_batched
    tensor_ops::upscale2d::tests::test_upscale2d_bilinear_even
    tensor_ops::upscale2d::tests::test_upscale2d_bilinear_uneven
    tensor_ops::upscale2d::tests::test_upscale2d_nearest_uneven
```

File: src/tensor_ops/concat/cuda_kernel.rs [hunk 4]
```diff
@@ -39,6 +39,7 @@ impl<E: Dtype + CudaTypeName> super::ConcatKernel<E> for Cuda {
             let src = BWD_KERNEL.replace("$Ty", E::NAME);
             let opts = CompileOptions {
                 arch: Some(env!("CUDA_COMPUTE_CAP")),
+                include_paths: vec!["/usr/include".to_string()],
```
Comment (coreylowman) [inline src/tensor_ops/concat/cuda_kernel.rs:42]: This is done a couple places to include "cuda_fp16.h". What is a better string for this? Could we use cuda toolkit root as include directory? Might require some updates to cudarc to set that env variable

**Comment by coreylowman** [comment]
Okay @nkoppel @opfromthestart, tests are compiling & running for both cuda & cpu. Both cpu & cuda have accuracy issues, perhaps more than I'd expect. For example cuda gelu might need to do computations in f32/f64 and then cast back to f16.

Here are cuda failures:
```
failures:
    losses::tests::test_bce_wide_range
    losses::tests::test_huber_loss
    losses::tests::test_mae
    losses::tests::test_smooth_l1_loss
    nn::batchnorm1d::tests::test_batchnorm1d_2d_forward_mut
    nn::batchnorm2d::tests::test_batchnorm2d_4d_forward_mut
    nn::linear::tests::test_forward_1d
    nn::linear::tests::test_linear_initialize
    nn::residual::tests::test_residual_gradients
    nn::unbiased_linear::tests::test_forward_1d
    optim::adam::tests::test_adam_decoupled_decay
    optim::adam::tests::test_custom_adam_one_params
    optim::adam::tests::test_default_adam_params
    optim::rmsprop::tests::test_rmsprop_centered
    optim::rmsprop::tests::test_rmsprop_decoupled_weight_decay
    optim::rmsprop::tests::test_rmsprop_default
    optim::rmsprop::tests::test_rmsprop_diff_alpha
    optim::rmsprop::tests::test_rmsprop_diff_eps
    optim::rmsprop::tests::test_rmsprop_l2_weight_decay
    optim::rmsprop::tests::test_rmsprop_momentum
    optim::sgd::tests::test_perfect_sgd
    tensor_ops::abs::tests::test_abs
    tensor_ops::broadcast_to::tests::test_broadcast_backwards
    tensor_ops::div::tests::test_div_2d
    tensor_ops::gelu::tests::test_gelu
    tensor_ops::log_softmax::tests::test_log_softmax_equivalence
    tensor_ops::matmul::tests::test_matmul_batched_3d
    tensor_ops::matmul::tests::test_matmul_batched_4d
    tensor_ops::matmul::tests::test_matmul_broadcast
    tensor_ops::matmul::tests::test_matmul_vec_normal
    tensor_ops::matmul::tests::test_matmul_vec_transpose
    tensor_ops::matmul::tests::test_small_matmul_mm
    tensor_ops::max_to::tests::test_max_axis_0_2d
    tensor_ops::max_to::tests::test_max_axis_1_2d
    tensor_ops::max_to::tests::test_max_negative_zero
    tensor_ops::mean_to::tests::test_mean_axis_0_2d
    tensor_ops::min_to::tests::test_min_axis_0_2d
    tensor_ops::min_to::tests::test_min_axis_1_2d
    tensor_ops::min_to::tests::test_min_negative_zero
    tensor_ops::pow::tests::test_powf_positive
    tensor_ops::pow::tests::test_powi_positive
    tensor_ops::softmax::tests::test_softmax_equivalence
    tensor_ops::sum_to::tests::test_sum_broadcasted
    tensor_ops::tanh::tests::test_tanh
    tensor_ops::upscale2d::tests::test_bilinear_upscale2d_batched
    tensor_ops::upscale2d::tests::test_upscale2d_bilinear_even
    tensor_ops::upscale2d::tests::test_upscale2d_bilinear_uneven
    tensor_ops::upscale2d::tests::test_upscale2d_nearest_uneven
```

We also need to implement atomicMaxf and atomicMinf for half (I left those as todo)

**Comment by ViliamVadocz** [comment]
> We also need to implement atomicMaxf and atomicMinf for half (I left those as todo)

I can do that.

I also want to take a look at how you did the kernels in case I see something that could be improved.

**Comment by ViliamVadocz** [comment]
An issue I am running into is that my GPU only has compute capability 6.1 so I am missing some functions such as `atomicSum` for `__half`. This will be an even bigger issue once we add `bfloat16` support. What should we do? Should we try to recreate the functionality with conversions if the compute capability is too low or should we disable those features?

**Comment by coreylowman** [comment]
Yeah bfloat16 requires cuda toolkit version 12+, so there's gonna be a lot of configuration issues I think.

How many functions are you missed on 6.1? If it's just atomicSum we can probably add that ourselves with some `#ifdef`.

Alternatively: for now we could raise compilation error in `build.rs` if we detect too low of a compute capability, and then address this in a future PR

**Comment by ViliamVadocz** [comment]
I am also missing some other functions which I left in comments in my initial merge request #705. Specifically, `__hmax_nan` and `__hmin_nan`.

**Comment by coreylowman** [comment]
Ahhh right. I mean since we have to add atomicMin/atomicMax anyway, I assume atomicSum would have a similar implementation? It seems straightforward enough to provide a fallback impl if the compute cap isn't high enough.

**Comment by ViliamVadocz** [comment]
We can always just convert to f32 in the worst case.

**Comment by coreylowman** [comment]
> We can always just convert to f32 in the worst case.

I like this idea, but the storage we'd be writing to is f16 I think, so is this possible?

File: src/tensor_ops/utilities/cuda_utils.cuh [hunk 5]
```diff
@@ -109,21 +117,43 @@ extern "C" __global__ void fill_with_f64(double *buf, double value, const size_t
...
+__device__ __forceinline__ __half powg(__half a, __half b) { return hexp(__hmul(hlog(a), b)); }
 __device__ __forceinline__ float tanhg(float a) { return tanhf(a); }
 __device__ __forceinline__ double tanhg(double a) { return tanh(a); }
+__device__ __forceinline__ __half tanhg(__half a) { return __hdiv(hsin(a), hcos(a)); }
 __device__ __forceinline__ float maxg(float a, float b) { return fmaxf(a, b); }
 __device__ __forceinline__ double maxg(double a, double b) { return fmax(a, b); }
+__device__ __forceinline__ __half maxg(__half a, __half b) { return __hmax(a, b); } // __hmax_nan
 __device__ __forceinline__ float ming(float a, float b) { return fminf(a, b); }
 __device__ __forceinline__ double ming(double a, double b) { return fmin(a, b); }
+__device__ __forceinline__ __half ming(__half a, __half b) { return __hmin(a, b); } // __hmin_nan
```
Comment (coreylowman) [inline src/tensor_ops/utilities/cuda_utils.cuh:147]: Note: figure out how to propagate nans

**Comment by coreylowman** [comment]
@ViliamVadocz I updated the include paths, let me know if the new commit passes for you

**Comment by coreylowman** [comment]
Pasting for later in build.rs:
```diff
+            let destination = std::format!(
+                "{}/{}",
+                out_dir,
+                path.file_name().unwrap().to_str().unwrap()
+            );
+            println!("cargo:rerun-if-changed={}", destination);
+            std::fs::copy(path.clone(), destination).unwrap();
```

**Comment by coreylowman** [comment]
@ViliamVadocz it seems like the atomicCAS or atomicAdd(half) is still broken  for old compute capabilities. I can take a look later today, but if you've got time feel free!

File: src/optim/adam/adam.cu [hunk 6]
```diff
@@ -56,11 +58,62 @@ __device__ void adam_update(
     param[i] -= g;
 }
 
+__device__ void adam_update(
```
Comment (coreylowman) [inline src/optim/adam/adam.cu:61]: Can remove this since we aren't doing this for Cpu. Pytorch also has accuracy issues if doing training in pure f16

File: src/optim/rmsprop/rmsprop.cu [hunk 7]
```diff
@@ -75,6 +76,73 @@ __device__ void rmsprop_update(
     param[i] -= g;
 }
 
+__device__ void rmsprop_update(
```
Comment (coreylowman) [inline src/optim/rmsprop/rmsprop.cu:79]: Can remove this for same reason as specialized adam_update

**Comment by ViliamVadocz** [comment]
> @ViliamVadocz it seems like the atomicCAS or atomicAdd(half) is still broken for old compute capabilities. I can take a look later today, but if you've got time feel free!

I'll take a look again.

> @ViliamVadocz I updated the include paths, let me know if the new commit passes for you

Will do!

**Comment by ViliamVadocz** [comment]
`__hmin` and `__hmax` are already defined on compute capability (CC) 6.1, so I had to change when they get added (`__CUDA_ARCH__ < 600` works). After that I managed to compile and run tests. Currently 247 passed; 110 failed.

**Comment by coreylowman** [comment]
On compute_cap 86 I'm only getting 8 failures

**Comment by coreylowman** [comment]
They are likely failing because sum is failing. I'm running `cargo +nightly test --tests -F test-f16,cuda sum` to run juts the sum tests

**Comment by ViliamVadocz** [comment]
In that case I'll just implement the atomicAdd directly instead of fiddling around with atomicCAS.  It does mean that compatibility code will leech into the min_to and max_to files since they also use atomicCAS on shorts.

**Comment by coreylowman** [comment]
Okay this is where I'm at now:
```c++
__device__ __half atomicAdd(__half* address, __half val) {
    size_t align = reinterpret_cast<size_t>(address) & 2;
    unsigned int *address_as_u32 = reinterpret_cast<unsigned int *>(reinterpret_cast<char *>(address) - align);
    unsigned int old = *address_as_u32;
    unsigned int assumed;

    do {
        assumed = old;
        __half sum16 = __ushort_as_half(align ? (old >> 16) : (old & 0xffff)) + val;
        unsigned int sum32 = (unsigned int) __half_as_ushort(sum16);
        old = align ? ((sum32 << 16) | (old & 0xffff)) : ((old & 0xffff0000) | sum32);
        old = atomicCAS(address_as_u32, assumed, old);
    } while (assumed != old);
    return __ushort_as_half(align ? (old >> 16) : (old & 0xffff));
}
```

375 passed, 18 failed.

It seems like this doesn't handle inf properly as some of the errors i'm still getting are:
```rust

---- tensor_ops::max_to::tests::test_max_axis_0_2d stdout ----
thread 'tensor_ops::max_to::tests::test_max_axis_0_2d' panicked at 'lhs != rhs | -inf != 3', src/tensor_ops/max_to/mod.rs:97:9

---- tensor_ops::max_to::tests::test_max_axis_1_2d stdout ----
thread 'tensor_ops::max_to::tests::test_max_axis_1_2d' panicked at 'lhs != rhs | -inf != 2', src/tensor_ops/max_to/mod.rs:112:9

---- tensor_ops::max_to::tests::test_max_negative_zero stdout ----
thread 'tensor_ops::max_to::tests::test_max_negative_zero' panicked at 'lhs != rhs | -inf != 0', src/tensor_ops/max_to/mod.rs:136:9

---- tensor_ops::min_to::tests::test_min_axis_0_2d stdout ----
thread 'tensor_ops::min_to::tests::test_min_axis_0_2d' panicked at 'lhs != rhs | inf != 1', src/tensor_ops/min_to/mod.rs:97:9

---- tensor_ops::min_to::tests::test_min_axis_1_2d stdout ----
thread 'tensor_ops::min_to::tests::test_min_axis_1_2d' panicked at 'lhs != rhs | inf != 1', src/tensor_ops/min_to/mod.rs:112:9

---- tensor_ops::min_to::tests::test_min_negative_zero stdout ----
thread 'tensor_ops::min_to::tests::test_min_negative_zero' panicked at 'lhs != rhs | inf != -0', src/tensor_ops/min_to/mod.rs:136:9
```

**Comment by ViliamVadocz** [comment]
All those failing tests are min and max which use the probably broken atomicCAS. I'm almost done with my attempt.

**Comment by ViliamVadocz** [comment]
PR #742 is up

**Comment by coreylowman** [comment]
@ViliamVadocz nice work, all the tests pass for me now! 🚀 (other than the ones I broke from reverting optimizer kernels)

File: src/optim/adam/mod.rs [hunk 8]
```diff
@@ -201,13 +203,15 @@ mod tests {
         let mut opt = Adam::new(
             &t,
             AdamConfig {
-                lr: 1e-3,
-                betas: [0.5, 0.25],
-                eps: 1e-8,
+                lr: NumCast::from(1e-3).unwrap(),
+                betas: [0.5, 0.25].map(NumCast::from).map(Option::unwrap),
+                eps: NumCast::from(1e-8).unwrap(),
```
Comment (coreylowman) [inline src/optim/adam/mod.rs:208]: Revert this

File: src/optim/adam/mod.rs [hunk 9]
```diff
@@ -231,12 +235,14 @@ mod tests {
...
+        let mut t = dev
+            .tensor([-0.5, -0.25, 0.1, 0.6, 1.0])
+            .to_dtype::<TestDtype>();
         let mut opt = Adam::new(
             &t,
             AdamConfig {
-                betas: [0.5, 0.25],
-                weight_decay: Some(WeightDecay::L2(1.0)),
+                betas: [0.5, 0.25].map(NumCast::from).map(Option::unwrap),
+                weight_decay: Some(WeightDecay::L2(NumCast::from(1.0).unwrap())),
```
Comment (coreylowman) [inline src/optim/adam/mod.rs:245]: revert these

File: src/optim/rmsprop/mod.rs [hunk 10]
```diff
@@ -201,9 +203,9 @@ mod tests {
     #[test]
     fn test_rmsprop_default() {
         let cfg = RMSpropConfig {
-            lr: 1e-2,
-            alpha: 0.9,
-            eps: 1e-8,
+            lr: NumCast::from(1e-2).unwrap(),
+            alpha: NumCast::from(0.9).unwrap(),
+            eps: NumCast::from(1e-8).unwrap(),
```
Comment (coreylowman) [inline src/optim/rmsprop/mod.rs:208]: revert this

File: src/optim/sgd/mod.rs [hunk 11]
```diff
@@ -206,7 +206,7 @@ mod tests {
         let mut sgd = Sgd::new(
             &pred,
             SgdConfig {
-                lr: 1.0,
+                lr: NumCast::from(1.0).unwrap(),
```
Comment (coreylowman) [inline src/optim/sgd/mod.rs:209]: revert

File: src/tensor_ops/max_to/max_to.cu [hunk 12]
```diff
@@ -1,5 +1,38 @@
 #include "cuda_utils.cuh"
 
+__device__ __forceinline__ __half atomicMaxf(__half* address, __half val) {
+#if __CUDA_ARCH__ < 700
```
Comment (coreylowman) [inline src/tensor_ops/max_to/max_to.cu:4]: Let's move this & the methods below into compatibility.cuh

File: src/tensor_ops/min_to/min_to.cu [hunk 13]
```diff
@@ -1,6 +1,39 @@
 #include "cuda_utils.cuh"
 
-// atomicMax is not implemented for floats,
+__device__ __forceinline__ __half atomicMinf(__half* address, __half val) {
+#if __CUDA_ARCH__ < 700
```
Comment (coreylowman) [inline src/tensor_ops/min_to/min_to.cu:4]: Same thing here move to compatibility.cuh
